src/osd/OSD.cc

```diff
@@ -9909,11 +9972,9 @@ void OSD::ShardedOpWQ::_process(uint32_t thread_index, heartbeat_handle_d *hb)
   osd->service.maybe_inject_dispatch_delay();
 
   // [lookup +] lock pg (if we have it)
-  if (!pg) {
-    pg = osd->_lookup_lock_pg(item.first);
-  } else {
-    pg->lock();
-  }
+  PGQueueable::OrderLocker::Ref l;
+  l = item.second.get_order_locker(pg, item.first, osd);
+  l->lock();
 
```

Comment:
@myoungwon yes, i agreed. we need to generalize the lock interface **and** the pg_slots side. i am working on it.